content/blog/single-message-transforms-swiss-army-knife-of-kafka-connect.asciidoc

```diff
@@ -0,0 +1,235 @@
+only exposing the new field name going forward.
+Similarly, a field that got removed from a message schema could be re-added,
+using some sort of constant placeholder value.
+Again consumers could then be updated at their own pace to not expect and access that field any longer.
+
+It should be said though that there are limits for this usage: e.g. when changing the type of a field,
+there's not much that could be done using an SMT for smoothening such transition.
+
+_Mac's rating:_ &nbsp;&nbsp; 📎📎📎&nbsp;&nbsp; _SMTs can help to address basic compatibility concerns around schema evolution_
+
+== Filtering and Routing
+
+When applied on the source side,
+SMTs allow to filter out specific records produced by the connector.
+They also can be used for controlling the Kafka topic a record gets sent to.
+That's in particular interesting when basing filtering and routing on the actual record contents.
+In an IoT scenario for instance where Kafka Connect is used to ingest data from some kind of sensors,
+an SMT might be used to filter out all sensor measurements _below_ a certain threshhold, or route measurement events _above_ a threshold to a special topic.
+
+Debezium  provides a range of SMTs for record filtering and routing:
+
+* The https://debezium.io/documentation/reference/configuration/topic-routing.html[logical topic routing SMT] allows to send change events originating from multiple tables to the same Kafka topic,
+which can be useful when working with partition tables in Postgres, or with data that is sharded into multiple tables
+* The https://debezium.io/documentation/reference/1.2/configuration/filtering.html[`Filter`] and `ContentBasedRouter` SMTs let you use script expressions in languages such as Groovy or JavaScript for filtering and routing change events based on their contents;
+such script-based approach can be an interesting middleground between ease-of-use (no Java code must be compiled and deployed to Kafka Connect) and expressiveness; e.g. here is how the routing SMT could be used with https://github.com/graalvm/graaljs[GraalVM's JavaScript engine] for routing change events from a table with purchase orders to different topics in Kafka, based on the order type:
++
+[source]
+----
+...
+transforms=route
+transforms.route.type=io.debezium.transforms.ContentBasedRouter
+transforms.route.topic.regex=.*purchaseorders
+transforms.route.language=jsr223.graal.js
+transforms.route.topic.expression=
+    value.after.ordertype == 'B2B' ? 'b2b_orders' : 'b2c_orders'
+...
+----
+* The https://debezium.io/documentation/reference/configuration/outbox-event-router.html[outbox event router] comes in handy when implementing the https://microservices.io/patterns/data/transactional-outbox.html[transactional outbox pattern] for data propagation between microservices: it can be used to send events originating from a single outbox table to a specific Kafka topic per aggregate or event type
```

Comment:
do you mean a DDD aggregate here? if so I'm sorry to tell you that you have no idea 😂

Reply:
😄 Ahaha.

Tbh. I completely did *not* think of that little episode when writing this. I think I won't elaborate any further on this. It's touched on in more depth in the SMT's docs.